### PR TITLE
[Merged by Bors] - Spark history server

### LINKED ISSUE #124: Implement ADR 22 (Spark history server)
## Description

Add the possibility to provision Spark history servers as described in [ADR022](https://docs.stackable.tech/home/stable/contributor/adr/ADR022-spark-history-server.html).
See: #84 

## Acceptance criteria

- [ ]  Can set up history servers and spark applications can log to them.
- [ ] Can clean up history servers.
- [ ] Documentation is updated.
- [ ] Tests are added.

### PR DESCRIPTION
**Pull request by razvan**
Fixes #124 #162
Jenkins: [Azure:heavy_check_mark:](https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/spark-k8s-operator-it-custom/60/)

### CODE REVIEW

File: docs/modules/ROOT/examples/example-history-server.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+---
+apiVersion: spark.stackable.tech/v1alpha1
+kind: SparkHistoryServer
+metadata:
+  name: spark-history
+spec:
+  image:
+    productVersion: 3.3.0
+    stackableVersion: 0.3.0
+  logFileDirectory:  # <1>
```
[the remark this answers is not in the document]
Comment (razvan) [reply docs/modules/ROOT/examples/example-history-server.yaml:10]: you can only have one cleaner

File: docs/modules/ROOT/pages/history_server.adoc [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+The example below demonstrates how to set up the history server running in one Pod with scheduled cleanups of the event logs. The event logs are loaded from an S3 bucket named `spark-logs` and the folder `eventlogs/`. The credentials for this bucket are provided by the secret class `s3-credentials-class`. For more details on how the Stackable Data Platform manages S3 resources see the xref:home:concepts:s3.adoc[S3 resources] page.
+
+
+[source,yaml]
+----
+include::example$example-history-server.yaml[]
+----
+
+<1> The location of the event logs. Must be a S3 bucket. Future implementations might add support for other shared filesystems such as HDFS.
+<2> Folder within the S3 bucket where the log files are located. This is folder is required and mus exist before setting up the history server.
```
Comment (adwk67) [inline docs/modules/ROOT/pages/history_server.adoc:18]:
```suggestion
<2> Folder within the S3 bucket where the log files are located. This folder is required and must exist before setting up the history server.
```

File: docs/modules/ROOT/pages/history_server.adoc [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+<3> The S3 bucket definition, here provided in-line.
+<4> Additional gistory server configuration properties can be provided here as a map. For possible properties see: https://spark.apache.org/docs/latest/monitoring.html#spark-history-server-configuration-options
+<5> This deployment has only one Pod. Multiple history servers can be started, all reading the same event logs by increasing the relica count.
+<6> This history server will automatically clean up old log files by using default properties. You can change any of these by using the `sparkConf` map.
+
+NOTE: Only one role group can have scheduled cleanups enabled (`cleaner: true`) and this role group can have a maximum replica of 1.
```
Comment (adwk67) [inline docs/modules/ROOT/pages/history_server.adoc:24]:
```suggestion
NOTE: Only one role group can have scheduled cleanups enabled (`cleaner: true`) and this role group cannot have more than 1 replica.
```

File: docs/modules/ROOT/pages/history_server.adoc [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+== Application configuration
+
+
+The example below demonstrates how to configure Spark applications store log events to a S3 bucket.
```
Comment (adwk67) [inline docs/modules/ROOT/pages/history_server.adoc:29]:
```suggestion
The example below demonstrates how to configure Spark applications to write log events to an S3 bucket.
```

File: docs/modules/ROOT/examples/example-history-server.yaml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+      bucket:  # <3>
+        inline:
+          bucketName: spark-logs
+          connection:
+            inline:
+              host: test-minio
+              port: 9000
+              accessStyle: Path
+              credentials:
+                secretClass: s3-credentials-class
```
Comment (adwk67) [inline docs/modules/ROOT/examples/example-history-server.yaml:22]: I would change the credentials class here from `s3-credentials-class` to `history-credentials-class`, like we have in the integration test, so it's more obvious that trying to use _those_ credentials in the SparkApplication app will just be ignored.
```suggestion
                secretClass: history-credentials-class
```

File: rust/crd/src/s3logdir.rs [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+        if let Some(conn) = s3bucket.as_ref().and_then(|i| i.connection.as_ref()) {
+            if let Some(tls) = &conn.tls {
+                match &tls.verification {
+                    TlsVerification::None {} => return S3TlsNoVerificationNotSupportedSnafu.fail(),
+                    TlsVerification::Server(server_verification) => {
+                        match &server_verification.ca_cert {
+                            CaCert::WebPki {} => {}
+                            CaCert::SecretClass(_) => {
+                                return S3TlsCaVerificationNotSupportedSnafu.fail()
```
Comment (sbernauer) [inline rust/crd/src/s3logdir.rs:83]: Any specific reason not add the ca.crt to the java truststore?

File: rust/crd/src/s3logdir.rs [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+                }
+            }
+        }
+
+        if let Some(conn) = s3bucket.as_ref().and_then(|i| i.connection.as_ref()) {
+            if conn.tls.as_ref().is_some() {
+                tracing::warn!("The resource indicates S3-access should use TLS: TLS-verification has not yet been implemented \
+            but an HTTPS-endpoint will be used!");
+            }
+        }
```
Comment (sbernauer) [inline rust/crd/src/s3logdir.rs:96]: I think this statement is not needed. All the possible cases are handled in the match statement above and are either executed correctly or an error is returned

File: rust/crd/src/s3logdir.rs [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    /// Constructs the properties needed for loading event logs from S3.
+    /// These properties are later written in the `HISTORY_CONFIG_FILE_NAME_FULL` file.
+    ///
+    /// The following properties related to credentials are not included:
+    /// * spark.hadoop.fs.s3a.aws.credentials.provider
+    /// * spark.hadoop.fs.s3a.access.key
+    /// * spark.hadoop.fs.s3a.secret.key
+    /// instead, the environment variables AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY are set
```
Comment (sbernauer) [inline rust/crd/src/s3logdir.rs:110]: Seems a bit weird to have different flows for spark applications and history server but fine for me ;)

File: rust/crd/src/s3logdir.rs [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+            }
+        }
+        result
+    }
+
+    pub fn application_spark_config(&self) -> BTreeMap<String, String> {
+        let mut result = BTreeMap::new();
+        result.insert("spark.eventLog.enabled".to_string(), "true".to_string());
+        result.insert("spark.eventLog.dir".to_string(), self.url());
+        result
```
Comment (sbernauer) [inline rust/crd/src/s3logdir.rs:136]: I think we need to configure endpoint, credentials, access style etc. here
Recall the ADR

> The operator will read the s3Connection and set up the fs.s3a.aws.credentials.provider and co (endpoint, accesskey, secretkey, tls, path-style access - basically all attributes of S3Connection) settings. Afterwards if logFileDirectory attribute is set: Set fs.s3a.bucket.<logFileDirectory-bucket-name(here spark-eventlogs)>.aws.credentials.provider and co to overwrite endpoint and credentials for the logging bucket. Set spark.eventLog.enabled property to true and will construct the spark.eventLog.dir from s3a://<logFileDirectory-bucket-name(here spark-eventlogs)>/<logFileDirectory-prefix(here eventlogs/)>.

File: rust/crd/src/s3logdir.rs [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+    }
+    pub fn credentials_volume(&self) -> Option<Volume> {
+        self.credentials()
+            .map(|credentials| credentials.to_volume(VOLUME_NAME_S3_CREDENTIALS))
+    }
+
+    pub fn credentials_volume_mount(&self) -> Option<VolumeMount> {
+        self.credentials().map(|_| VolumeMount {
+            name: VOLUME_NAME_S3_CREDENTIALS.into(),
+            mount_path: S3_SECRET_DIR_NAME.into(),
```
Comment (sbernauer) [inline rust/crd/src/s3logdir.rs:154]: Please keep in mind that we can have multiple s3 secrets (one for application and one for eventlogs). This is the reason i generally name volume name and volume mount path `<secret-class-name>-s3-credentials`

File: tests/templates/kuttl/spark-history-server/10-deploy-spark-app.yaml.j2 [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+      bucket: # S3BucketDef
+        inline:
+          bucketName: spark-logs
+          connection:
+            inline:
+              host: test-minio
+              port: 9000
+              accessStyle: Path
+              credentials:
+                secretClass: history-credentials-class
```
Comment (sbernauer) [inline tests/templates/kuttl/spark-history-server/10-deploy-spark-app.yaml.j2:35]: Can we please have a test using different endpoints for data payload and eventlogs? This test should uncover the previously mentioned problem of currently not setting endpoint and credentials specific for the eventlog bucket.

**Review by sbernauer — COMMENTED** [review]
Not a full review, as this issue is still in Progress column.
I just wanted to address some problems I spotted early on.
Please feel free to ignore them until actual Review starts, i can also offer to do a "full" review ;)

File: rust/crd/src/s3logdir.rs (hunk 10 above)
Comment (razvan) [reply rust/crd/src/s3logdir.rs]: Done.

File: rust/crd/src/s3logdir.rs (hunk 7 above)
Comment (razvan) [reply rust/crd/src/s3logdir.rs]: Done.

File: rust/crd/src/s3logdir.rs [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+    fn url(&self) -> String {
+        format!(
+            "s3a://{}/{}",
+            self.bucket.bucket_name.as_ref().unwrap().clone(), // this is guaranteed to exist at this point
+            self.prefix
+        )
+    }
+    pub fn credentials_volume(&self) -> Option<Volume> {
+        self.credentials()
+            .map(|credentials| credentials.to_volume(credentials.secret_class.as_ref()))
```
Comment (sbernauer) [inline rust/crd/src/s3logdir.rs:181]: I personally would prefer calling the volume and volume mount something like `<secret_class>-s3-credentials` for better debugging as well as being absolutely sure no paths will clash.
Feel free to ignore this as it's a minor detail

File: rust/crd/src/s3logdir.rs (hunk 9 above)
Comment (razvan) [reply rust/crd/src/s3logdir.rs:169]: Done.

File: tests/templates/kuttl/spark-history-server/10-deploy-spark-app.yaml.j2 (hunk 11 above)
Comment (razvan) [reply tests/templates/kuttl/spark-history-server/10-deploy-spark-app.yaml.j2:35]: Done.

File: rust/crd/src/s3logdir.rs (hunk 12 above)
Comment (razvan) [reply rust/crd/src/s3logdir.rs:181]: Done.

**Review by sbernauer — CHANGES_REQUESTED** [review]
Haven't gotten to Rust code yet, leaving early feedback

File: docs/modules/ROOT/pages/history_server.adoc [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+<3> Instruct the operator to configure the application with logging enabled.
+<4> Folder to store logs. This must match the prefix used by the history server.
+<5> Bucket to store logs. This must match the bucket used by the history server.
+<6> Credentials used to write event logs. These can, of course, differ from the credentials used to process data.
+
+
+
+== History Web UI
+
+The history exposes a user console on port 18080. By setting up port-forwarding on 18080 this UI can be opened in a browser to show running and completed jobs:
```
[the remark this answers is not in the document]
Comment (razvan) [reply docs/modules/ROOT/pages/history_server.adoc]: Services are being deployed already. The docs are updated.

File: deploy/helm/spark-k8s-operator/templates/deployment.yaml [hunk 14]
```diff
@@ -31,6 +31,9 @@ spec:
         {{- toYaml .Values.podSecurityContext | nindent 8 }}
       containers:
         - name: {{ include "operator.appname" . }}
+          env:
```
[the remark this answers is not in the document]
Comment (razvan) [reply deploy/helm/spark-k8s-operator/templates/deployment.yaml]: Removed

File: Makefile [hunk 15]
```diff
@@ -23,6 +23,10 @@ SHELL=/usr/bin/env bash -euo pipefail
 render-readme:
 	scripts/render_readme.sh
 
+## Alternative Dockerfile that uses cargo chef to speed up dev builds.
```
[the remark this answers is not in the document]
Comment (razvan) [reply Makefile]: Removed

File: CHANGELOG.md [hunk 16]
```diff
@@ -4,6 +4,13 @@ All notable changes to this project will be documented in this file.
 
 ## [Unreleased]
 
+### Added
+
+- Create and manage history servers ([#187])
+- `operator-rs` `0.27.1` -> `0.30.2` ([#187])
```
[the remark this answers is not in the document]
Comment (razvan) [reply CHANGELOG.md]: Done.

File: docs/modules/ROOT/examples/example-history-app.yaml [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+metadata:
+  name: spark-pi-s3-1
+spec:
+  version: "1.0"
+  sparkImage: docker.stackable.tech/stackable/spark-k8s:3.3.0-stackable0.3.0
+  sparkImagePullPolicy: IfNotPresent
+  mode: cluster
+  mainClass: org.apache.spark.examples.SparkPi
+  mainApplicationFile: s3a://my-bucket/spark-examples_2.12-3.3.0.jar
+  s3connection: # <1>
```
[the remark this answers is not in the document]
Comment (razvan) [reply docs/modules/ROOT/examples/example-history-app.yaml:13]: Done for the kuttl tests.

File: rust/operator-binary/src/history_controller.rs [hunk 18]
```diff
@@ -0,0 +1,511 @@
...
+    let empty = BTreeMap::new();
+
+    let mut log_dir_settings = s3_log_dir.history_server_spark_config();
+
+    // add cleaner spark settings if requested
+    log_dir_settings.extend(cleaner_config(shs, rolegroupref)?);
+
+    // add user provided configuration. These can overwrite everything.
+    let user_settings = shs.spec.spark_conf.as_ref().unwrap_or(&empty);
+    log_dir_settings.extend(user_settings.clone().into_iter());
```
Comment (sbernauer) [inline rust/operator-binary/src/history_controller.rs:432]: `empty` can be removed
```suggestion
    log_dir_settings.extend(shs.spec.spark_conf.clone().unwrap_or_default());
```

File: rust/operator-binary/src/spark_k8s_controller.rs [hunk 19]
```diff
@@ -87,6 +87,10 @@ pub enum Error {
         source: stackable_operator::error::Error,
         container_name: String,
     },
+    #[snafu(display("failed to resolve the s3 log dir confirguration"))]
```
Comment (sbernauer) [inline rust/operator-binary/src/spark_k8s_controller.rs:90]:
```suggestion
    #[snafu(display("failed to resolve the s3 log dir configuration"))]
```

**Review by sbernauer — COMMENTED** [review]
Review finished. I would like to try it out in the meantime

File: tests/templates/kuttl/spark-history-server/10-deploy-spark-app.yaml.j2 (hunk 11 above)
Comment (sbernauer) [reply tests/templates/kuttl/spark-history-server/10-deploy-spark-app.yaml.j2]: Many thanks!

File: rust/crd/src/lib.rs [hunk 20]
```diff
@@ -393,17 +415,19 @@ impl SparkApplication {
                 None => {}
             }
             if conn.credentials.as_ref().is_some() {
```
[the remark this answers is not in the document]
Comment (razvan) [reply rust/crd/src/lib.rs]: Done

File: rust/operator-binary/src/pod_driver_controller.rs [hunk 21]
```diff
@@ -1,14 +1,14 @@
...
 use std::sync::Arc;
 use std::time::Duration;
 
 use snafu::{OptionExt, ResultExt, Snafu};
 use stackable_operator::logging::controller::ReconcilerError;
 use strum::{EnumDiscriminants, IntoStaticStr};
 
-pub const POD_DRIVER_CONTROLLER_NAME: &str = "pod-driver";
-
 const LABEL_NAME_INSTANCE: &str = "app.kubernetes.io/instance";
```
[the remark this answers is not in the document]
Comment (razvan) [reply rust/operator-binary/src/pod_driver_controller.rs:12]: Done.

**Review by sbernauer — APPROVED** [review]
LGTM
During review we noticed that we currently can't verify S3 connections using a SecretClass. Ideally this would be fixed in this PR, but I'm fine with releasing without it and working on the Issue https://github.com/stackabletech/spark-k8s-operator/issues/194 directly afterwards

**Comment by razvan** [comment]
bors merge

**Pull request closed without merging**